### ISSUE TITLE
Image | Add support for ARMv8 VMs

### ISSUE THREAD

**Issue by CypZ** [MichaIng/DietPi#5637]
### Creating an image request

Goal is to run DietPi on my rpi 4B but on top of ESXi, not on bare metal. As far as I can see only x86_64 images are available for VMware.

#### Formal device information
- Device name | VMware/ESXi ARMv8 64-bit
- Official product URL | https://flings.vmware.com/esxi-arm-edition
- GitHub resource URL (if available) | 
- Image download URL | 

#### Is the SBC officially supported by the Debian installer?
- Debian generally support arm64 (and VMware virtual environment too) : https://www.debian.org/releases/stretch/arm64/ch02s01.html.en

#### If not, is a reliable 3rd party Debian image available for this SBC?
- ...

#### If not, are there install instructions for Debian available?
- https://www.debian.org/releases/stretch/arm64/ch02s01.html.en
- https://williamlam.com/2020/10/how-to-run-raspberry-pi-os-as-a-vm-on-esxi-arm.html
- https://www.youtube.com/watch?v=JN1h2ztUHLo

**Comment by MichaIng** [comment]
@bbugh
As you requested the same, moving discussion here.

Providing a VMX appliance and VMware/ESXi can go hand in hand, I think there is not much needed to change within the VMX (config) file. For the containing image, AFAIK it is only `grub-efi-arm64` and the regular `linux-image-arm64` kernel? I guess we could also go with the `linux-image-current-arm64` kernel from Armbian, but aside of some lower disk space usage, I don't think that it has much benefits.

Tasks to do:
- Decide whether to add a new hardware ID for ARMv8 VMs oder whether to review and rework some parts of our code where ID 20 is currently treated as `x86_64`. Not sure what makes more work, I'd probably check of `$G_HW_MODEL == 20 ` in our code and see how often this happens and how much work it would be to add another differentiation for `$G_HW_ARCH` in case. But I think in most cases where the architecture plays a role we do already check for the architecture, while VMs of any architecture have sufficiently much in common to handle them below the same ID.
- Generate bare `img` (or VMDK) and import it into new ESXi-Arm VM. Then export the appliance if it works. What I want to know is the VMX and/or OVF config file, to compare with the `x86_64` ones and have a template. I hope that they are mostly identical.

**Comment by oxoocoffee** [comment]
Please let me know if I can assist by providing information or testing on MBP M1 Parallels 19 or UTM.

**Comment by Najihel** [comment]
Hello,

Same request to host dietpi VM on Freebox Server (KVM)

Thanks

**Comment by MichaIng** [comment]
They are, but no ETA. But I recently did some changes which make an implementation easier: https://github.com/MichaIng/DietPi/commit/f8002eb

**Comment by MichaIng** [comment]
Here is an image to test: https://dietpi.com/downloads/images/testing/
RAW format, hence depending on virtualizer, you might need to use qemu-img to convert it to VMDK, QCOW2,VHDX or such.

If it generally works, and someone can provide an OVA (at best the embedded OVF config only), or VMX config for a best default VM appliance, we can provide a respective appliance for that virtualizer as well.

**Comment by Maaaash32** [comment]
I successfully tested the ARMv8 DietPi image on a Freebox Delta virtual machine.

Virtualization setup:
- Freebox Delta
- QEMU 4.1.0
- EDK II firmware 201811
- ARM64/aarch64 guest
- 2 virtual CPUs
- Approximately 1 GiB RAM currently assigned to the VM
- UEFI boot
- QCOW2 disk, converted from the original RAW image

The VM boots successfully and is usable over SSH.

Guest information:
- Debian GNU/Linux 13 (trixie)
- Kernel: 6.12.101+deb13-arm64
- CPU model detected as ARM Cortex-A72
- UEFI detected successfully
- Root filesystem: ext4
- Disk detected as VirtIO block device (/dev/vda)
- Network detected as VirtIO and working correctly
- IPv4 networking works correctly

The original image had limited available disk space. After converting it to QCOW2, I enlarged the virtual disk manually and then expanded the Linux partition and ext4 filesystem. The VM now sees approximately 16 GiB of usable root filesystem space.

Relevant output:

    Architecture: aarch64
    CPU model: Cortex-A72
    UEFI: EFI v2.7 by EDK II
    Disk: VirtIO block device, /dev/vda
    Root filesystem: /dev/vda2, ext4, approximately 16 GiB
    Network: VirtIO Ethernet, working correctly

The kernel log shows successful initialization of EFI, VirtIO block storage, VirtIO SCSI, VirtIO networking and ext4.

I do not have an OVA, OVF or VMX file. The Freebox VM currently consists of a QCOW2 virtual disk and a UEFI variables/NVRAM file (.efivars). However, the image works correctly with this QEMU/EDK II ARM64 configuration.

[DietPi VM on Freebox Delta.txt](https://github.com/user-attachments/files/31346981/DietPi.VM.on.Freebox.Delta.txt)

**Comment by MichaIng** [comment]
Thanks! So images in QCOW2 format are definitely valuable. Is there any sort of appliance that works with this virtualiser, i.e. some format it can import/export, as template to speed up or skip the VM configuration?

> The Freebox VM currently consists of a QCOW2 virtual disk and a UEFI variables/NVRAM file (.efivars). However, the image works correctly with this QEMU/EDK II ARM64 configuration.

Where is the VM configuration itself stored, like the virtual devices etc (if not just CLI args)?

**Comment by Maaaash32** [comment]
I'm from France. A Freebox is an internet provider box very permissive that can have VM inside. https://www.free.fr/freebox/freebox-ultra
There are two models that can record TV streams and have VM inside : Delta and Ultra. https://www.lafibreoptique.fr/comparer-les-offres/freebox-ultra-vs-delta
Delta can have 4 SATA hard drives 2,5" and Ultra that can only have an SSD m2 drive inside. Both have 2GB of RAM and 2 cores dedicated for a maximum of 2 VM. The hardware is pretty much the same. Here is a guide someone made https://freebox.laurepro.fr/?pihole 

We have 3 ways to add a VM : choose an OS in a list provided (debian 12, ubuntu 24.04, Alpine 3.20), select an ISO to install or add a .qcow2 file.

These Freebox have a graphical interface Freebox OS on the ip of the box to configure it. You can have my screenshots.

<img width="2560" height="1278" alt="Image" src="https://github.com/user-attachments/assets/8a61d7c6-8d15-4627-96e3-cf1e2d1c927d" />

<img width="2560" height="1278" alt="Image" src="https://github.com/user-attachments/assets/127e8c64-548b-4e4b-a9e5-bedcb05856c7" />

<img width="2560" height="1278" alt="Image" src="https://github.com/user-attachments/assets/66b0e818-870c-468c-b2da-ffc72f3a52ee" />

<img width="2560" height="1278" alt="Image" src="https://github.com/user-attachments/assets/9658900a-bfee-47e5-a91f-2cbb681def50" />

I don't know where i can find the VM configuration files. I can only interact with it from console and SSH.

**Comment by MichaIng** [comment]
On your 3rd screenshot, when you select this "Add a VM" button at the left side, there is no way to import a VM, but just a dialog to confirm or adjust settings, like selecting a disk image, changing the number of CPUs (default 1) and RAM size (default 1024 MiB)?

**Comment by Maaaash32** [comment]
When I click on "Add a VM" button it's a dialog to create a VM with 3 options :
- select an OS from the list (debian 12, ubuntu 24.04, Alpine 3.20) this guide show it https://freebox.laurepro.fr/?alpine
- provide an ISO to install it
- provide a .qcow2 image

Then I choose the number of CPU and the amount of RAM.

But after the VM is created, I can modify the amount of RAM and the number of CPU on the right when it's turn off.

And on all cases, the VM is installed on a .qcow2 disk like you can see it on the 4th screenshot.

**Comment by MichaIng** [comment]
Okay, seems sufficiently easy with a QCOW2 then.

Installer ISOs for ARMv8 would be a topic in its own. Currently we use Clonezilla Love, with is x86_64 only.

**Comment by Maaaash32** [comment]
> The original image had limited available disk space. After converting it to QCOW2, I enlarged the virtual disk manually and then expanded the Linux partition and ext4 filesystem. The VM now sees approximately 16 GiB of usable root filesystem space.

I remind you that I had to do this. With the raw image converted in .qcow2 with qemu-img . At first I didn't know how to do it.
With the future .qcow2 image you'll provide, will I still have to do it ?

**Comment by MichaIng** [comment]
Right, this won't be needed anymore. I further integrated this into our build scripts, running a test build with a full set of ARMv8 Trixie VM images: https://github.com/MichaIng/DietPi/actions/runs/33407700538

This includes raw `.img` images, `.qcow2`, tarballs with VMX/VMDK appliances, ESXi+VirtualBox OVA appliances, an UTM appliance, and a VHDX/Hyper-V image. Not sure whether Hyper-V and VirtualBox run on Windows for ARM, and whether the VM configurations in the appliances are all compatible with the ARM VMs. If not, let me know, and we can adjust the configs for ARM or skip individual builds.

The images can be found here once done: https://dietpi.com/downloads/images/testing/

**Comment by Maaaash32** [comment]
I have a problem with your last .qcow2 image. The VM didn't give itself an IP adress. So I cannot connect to it via SSH and can't interact with it. The console inside freebox OS is only a screen, I can't write inside.
The raw image of june didn't had this IP adress problem. From the start, It had an IP adress so I could interact with it via SSH.

I can't give you a log so here is a screenshot

<img width="2560" height="1278" alt="Image" src="https://github.com/user-attachments/assets/d8bb6a72-d509-4fee-bc5a-72aaff5d44dc" />

**Comment by MichaIng** [comment]
Is the screenshot from the first boot, or did you reboot once already? Because I see no `dietpi-firstboot.service` running. But also possible that something went wrong with the new `dietpi-network` interface configuration. I'll review/test.

**Comment by Maaaash32** [comment]
No, I reboot it once already because it hasn't an IP on the first boot.

**Comment by MichaIng** [comment]
Found it, I missed to use `dev` branch for the DietPi code on the image, but did use the `dev` branch of the `dietpi-installer`. That way `/etc/network/interfaces` was empty, which is fine for v10.7, but not for v10.6.

New build running: https://github.com/MichaIng/DietPi/actions/runs/34512886036

**Comment by Maaaash32** [comment]
Hello, your last .qcow2 image for an ARMv8 VM of Dietpi 10.7 works like a charm on my Freebox Delta. 👍
Thank you very much ! 🥳

**Comment by MichaIng** [comment]
Great, many thanks for testing it again 🙂.
